mavros_extras/src/plugins/log_transfer.cpp

```diff
@@ -0,0 +1,125 @@
+#include <mavros/mavros_plugin.h>
+#include <mavros_msgs/LogData.h>
+#include <mavros_msgs/LogEntry.h>
+#include <mavros_msgs/LogRequestData.h>
+#include <mavros_msgs/LogRequestEnd.h>
+#include <mavros_msgs/LogRequestList.h>
+
+namespace mavros {
+namespace extra_plugins {
+class LogTransferPlugin : public plugin::PluginBase {
+public:
+	LogTransferPlugin() :
+		nh("~log_transfer") {}
+
+	void initialize(UAS& uas) override
+	{
+		PluginBase::initialize(uas);
+
+		log_entry_pub = nh.advertise<mavros_msgs::LogEntry>("raw/log_entry", 1000);
+		log_data_pub = nh.advertise<mavros_msgs::LogData>("raw/log_data", 1000);
+
+		log_request_list_srv = nh.advertiseService("raw/log_request_list",
+					&LogTransferPlugin::log_request_list_cb, this);
+		log_request_data_srv = nh.advertiseService("raw/log_request_data",
+					&LogTransferPlugin::log_request_data_cb, this);
+		log_request_end_srv = nh.advertiseService("raw/log_request_end",
+					&LogTransferPlugin::log_request_end_cb, this);
+	}
+
+	Subscriptions get_subscriptions() override
+	{
+		return {
+			       make_handler(&LogTransferPlugin::handle_log_entry),
+			       make_handler(&LogTransferPlugin::handle_log_data),
+		};
+	}
+
+private:
+	ros::NodeHandle nh;
+	ros::Publisher log_entry_pub, log_data_pub;
+	ros::ServiceServer log_request_list_srv, log_request_data_srv, log_request_end_srv;
+
+	void handle_log_entry(const mavlink::mavlink_message_t*, mavlink::common::msg::LOG_ENTRY& le)
+	{
+		mavros_msgs::LogEntry msg;
+		msg.id = le.id;
+		msg.num_logs = le.num_logs;
+		msg.last_log_num = le.last_log_num;
+		msg.time_utc = ros::Time(le.time_utc);
+		msg.size = le.size;
+		log_entry_pub.publish(msg);
+	}
+
+	void handle_log_data(const mavlink::mavlink_message_t*, mavlink::common::msg::LOG_DATA& ld)
+	{
+		mavros_msgs::LogData msg;
+		msg.id = ld.id;
+		msg.offset = ld.ofs;
+
+		decltype(ld.data) ::size_type count(ld.count);
+		if (count > ld.data.max_size()) {
+			count = ld.data.max_size();
+		}
+		msg.data.insert(msg.data.cbegin(), ld.data.cbegin(), ld.data.cbegin() + count);
```

Comment:
Why not `std::copy`?

Reply:
because then I'll have to resize the destination as a separate step: msg.data is a std::vector which is initialized to be of zero size upon construction. std::vector::insert does this for me